src/third/react-native/KeyboardAvoidingView.js

```diff
@@ -0,0 +1,251 @@
+import type { KeyboardEvent } from 'react-native/Libraries/Components/Keyboard/Keyboard';
+import type {
+  ViewProps,
+  ViewLayout,
+  ViewLayoutEvent,
+} from 'react-native/Libraries/Components/View/ViewPropTypes';
+
+type Props = $ReadOnly<{|
+  ...ViewProps,
+
+  /**
+   * Specify how to react to the presence of the keyboard.
+   */
+  behavior?: ?('height' | 'position' | 'padding'),
+
+  /**
+   * Style of the content container when `behavior` is 'position'.
+   */
+  contentContainerStyle?: ?ViewStyleProp,
+
+  /**
+   * Controls whether this `KeyboardAvoidingView` instance should take effect.
+   * This is useful when more than one is on the screen. Defaults to true.
+   */
+  enabled: ?boolean,
+
+  /**
+   * Distance between the top of the user screen and the React Native view. This
+   * may be non-zero in some cases. Defaults to 0.
+   */
+  keyboardVerticalOffset: number,
+|}>;
+
+type State = {|
+  bottom: number,
+|};
+
+/**
+ * View that moves out of the way when the keyboard appears by automatically
+ * adjusting its height, position, or bottom padding.
+ */
+class KeyboardAvoidingView extends React.Component<Props, State> {
+  static defaultProps: {|enabled: boolean, keyboardVerticalOffset: number|} = {
+    enabled: true,
+    keyboardVerticalOffset: 0,
+  };
+
+  _frame: ?ViewLayout = null;
+  _keyboardEvent: ?KeyboardEvent = null;
+  _subscriptions: Array<EventSubscription> = [];
+  // $FlowFixMe[unclear-type]
+  viewRef: {current: ElementRef<any> | null, ...};
+  _initialFrameHeight: number = 0;
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {bottom: 0};
+    this.viewRef = React.createRef();
+  }
+
+  _relativeKeyboardHeight(keyboardFrame): number {
+    const frame = this._frame;
+    if (!frame || !keyboardFrame) {
+      return 0;
+    }
+
+    if (keyboardFrame.height === 0) {
+        // Empirically, in default environments, the keyboard animation is
+        // slide-in and slide-out. The off-screen keyboard frame is a
+        // downward translation of the on-screen frame, to just off the
+        // screen, with no change in height or width.
+        //
+        // Empirically, if the UIAccessibilityPrefersCrossFadeTransitions
+        // setting is set (see
+        // https://developer.apple.com/documentation/uikit/3584818-uiaccessibilitypreferscrossfadet?language=objc)
+        // ), then the keyboard fades in and out instead of sliding. In the
+        // off-screen state, the keyboard-frame values (`height`, `width`,
+        // `screenX`, and `screenY`) all take the special value of zero.
+        //
+        // So, recognize that special representation as a keyboard-closed
+        // state. (We choose `height` for the conditional because it seems
+        // least likely to collide with some valid non-off-screen state.)
+        // See zulip/zulip-mobile#5162 and facebook/react-native#29974.
+        return 0;
+    }
```

Comment:
I think this can explain much more directly why this logic is correct. In particular the correctness of it doesn't depend on those empirical observations (always a bit concerning of a thing to have to rely on) or on very much detail at all.

Then the explanation of why it's needed comes after that. The possibility that it's redundant is lower-stakes than the possibility that it's incorrect, so the reader can be more relaxed about it.

Here's one version:
```suggestion
    if (keyboardFrame.height === 0) {
      // The keyboard is hidden, and occupies no height.
      //
      // This condition is needed because in some circumstances when the
      // keyboard is hidden we get both `screenY` and `height` (as well as
      // `screenX` and `width`) of zero.  In particular, this happens when
      // the UIAccessibilityPrefersCrossFadeTransitions setting is true,
      // i.e. when the user has enabled both "Reduce Motion" and "Prefer
      // Cross-Fade Transitions" under Settings > Accessibility > Motion.
      // See zulip/zulip-mobile#5162 and facebook/react-native#29974.
    }
```

(Another thing there is that I think saying how to find the setting in the UI is more helpful than that API docs page; we don't want to have any code looking for that setting in any case, at least not in connection with this logic. The page is easy to find via Google given the identifier, anyway.)

Reply:
Oh, and this component is written to be cross-platform, right? We only use it on iOS, but the conditional in `componentDidMount` is meant to operate on Android too.

So this description of the relevant circumstance should mention iOS. Adding "on iOS" after "In particular, this happens" should do it.